ansible/roles/riakcs/configure/defaults/main.yml

```diff
@@ -1,15 +1,39 @@
-# variables used in conf files
+# Possible values for riak_backend
+#
+# - "leveldb"
+#   The default, use this unless you know you need something different.
+#   Requires less RAM, lower performance.
+#
+# - "bitcask"
+#   All keys must fit in RAM. Higher performance, but requires more RAM. Uses
+#   a LOT more disk space possibly due to bitcask not using compression.
+#
+# - "leveldb-bad-config"
+#   Bad leveldb configuration. We're slowly migrating away from this, but the
+#   migration requires reading each object from the old cluster and saving
+#   it in a new cluster.
+#   http://lists.basho.com/pipermail/riak-users_lists.basho.com/2017-March/019142.html
+#   This configuration uses by far the least disk space. Would be cool if a
+#   change could be made to riak to support this configuration and eliminate
+#   the (low) chance of data loss that it apparently has.
+#
+riak_backend: "leveldb"
 
-# the following are defined in group_vars/all.yml
-#riak_data_dir: ...
-#riak_port: ...
-#riakcs_port: ...
-#stanchion_data_dir: ...
-#stanchion_port: ...
+# `riakcs_port` is defined in group_vars/all.yml
 
+riak_port: 8087
 riak_data_root_leveldb: "{{ riak_data_dir }}/leveldb"
+riak_data_root_bitcask: "{{ riak_data_dir }}/bitcask"
+riak_data_root_blocks: "{{ riak_data_dir }}/blocks"
+riak_data_dir: "{{ encrypted_root }}/riak"
 riak_ring_dir: "{{ riak_data_dir }}/ring"
-riak_nodename: "riak-{{ inventory_hostname_short }}@{{ inventory_hostname }}"
-riak_ring_size: 64
+
+# This parameter cannot be changed once the cluster has been configured and
+# has data in it. Use a lower ring size (64) in envs where RAM is very limited
+# since higher ring size requires more RAM, even for leveldb-only backend.
+riak_ring_size: 128
```

Comment:
We generally default to running our sevices on low hardware unless we really have to even if it impacts performance (not saying I agree to this, just mentioning it's a recurring theme)

What's the reason to default to 128 instead of 64? originally I thought there was going to be a point where our cluster could no longer grow, but I think that this config value just impacts the performance. Is that correct?

Reply:
It's a somewhat grey area. The [Riak documentation](http://docs.basho.com/riak/kv/2.2.3/setup/planning/cluster-capacity/#ring-size-number-of-partitions) recommends ring size >64 for cluster sizes >6 nodes. However, I've gleaned from talking to folks at Basho and discussions on the mailing list that the main concern of too low a ring size may be more about efficient resource usage (using maximal amounts of RAM) on very beefy hosts than it is about absolute ability to run. This of course assumes the cluster will always have less than 64 nodes (probably less than 32 or even 16, practically speaking) since 64 would be an absolute maximum number of nodes for a ring size of 64.

I would prefer to default to ring size of 128 since that allows clusters to grow much bigger before something drastic, like complete migration to a new cluster with a different ring size, needs to be done. As we know, that's a painful process.

As for running our clusters with very lean resources, I think this is really inefficient and a bad trade-off because of the number of man hours required to troubleshoot silly resource issues that should not even be an issue, but that's a discussion for a different place.

Reply:
Thanks for the explanation of ring size. Definitely makes sense to default to 128.

> As for running our clusters with very lean resources, I think this is really inefficient and a bad trade-off because of the number of man hours required to troubleshoot silly resource issues that should not even be an issue, but that's a discussion for a different place.

agreed on all points

Reply:
Incidentally, I've noticed that a ring size of 128 with cluster size of 6 nodes seems to be problematic.  Apparently the default ring balancing algorithm does not like that combination. During recent cluster resizing on prod I had a really hard time getting a plan that didn't include "WARNING: Not all replicas will be on distinct nodes," and there is one node (hqriak20) that has consistently higher CPU usage than the others. After many clear/plan iterations I thought I got (and applied) a plan that didn't have that warning, but later troubleshooting has made me suspect that something is still wrong. Unfortunately the diagnostic tools built into riak are broken on the version of riak we're using, so I'm not certain about any of this.